pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+		"ulong", "unchecked", "unsafe", "ushort",
+		"using", "virtual", "void", "volatile", "while":
+		return "@" + s
+
+	default:
+		return s
+	}
+}
+
+func csharpPropertyName(v *variable) string {
+	if v.info != nil && v.info.CSharpName != "" {
+		return v.info.CSharpName
+	}
+	return strings.Title(v.name)
+}
+
+// csharpPropertyType returns the C# type name for a given schema property.
+//
+// wrapInput can be set to true to cause the generated type to be deeply wrapped with `Input<T>`.
+func csharpPropertyType(v *variable, wrapInput bool) string {
+	return qualifiedCSharpPropertyType(v, "", wrapInput)
+}
+
+func isImmutableArrayType(typ *propertyType, wrapInput bool) bool {
+	return (typ.kind == kindSet || typ.kind == kindList) && !wrapInput
+}
+
+func qualifiedCSharpPropertyType(v *variable, qualifier string, wrapInput bool) string {
+	t := qualifiedCSharpType(v.typ, qualifier, wrapInput)
+	if v.optional() && !isImmutableArrayType(v.typ, wrapInput) {
+		t += "?"
+	}
+	return t
+}
+
+// qualifiedCsharpType returns the C# type name for a given schema value type and element kind.
+func qualifiedCSharpType(typ *propertyType, qualifier string, wrapInput bool) string {
+	// Prefer overrides over the underlying type.
+	var t string
+	switch {
+	case typ == nil:
+		return "object"
+	case typ.asset != nil:
+		if typ.asset.IsArchive() {
+			t = typ.asset.Type()
+		} else {
+			t = "AssetOrArchive"
+		}
+	default:
+		// First figure out the raw type.
+		switch typ.kind {
+		case kindBool:
+			t = "bool"
+		case kindInt:
+			t = "int"
+		case kindFloat:
+			t = "double"
+		case kindString:
+			t = "string"
+		case kindSet, kindList:
+			elemType := qualifiedCSharpType(typ.element, qualifier, false)
+			if wrapInput {
+				return fmt.Sprintf("InputList<%v>", elemType)
+			}
+			return fmt.Sprintf("ImmutableArray<%v>", elemType)
+		case kindMap:
+			elemType := qualifiedCSharpType(typ.element, qualifier, false)
+			if wrapInput {
+				return fmt.Sprintf("InputMap<%v>", elemType)
+			}
+			return fmt.Sprintf("ImmutableDictionary<string, %v>", elemType)
+		case kindObject:
+			t = typ.name
+			if qualifier != "" {
+				t = qualifier + "." + t
+			}
+		default:
+			contract.Failf("Unrecognized schema type: %v", typ.kind)
+		}
+	}
+
+	if wrapInput {
+		t = fmt.Sprintf("Input<%v>", t)
+	}
+
+	return t
+}
+
+func csharpPrimitiveValue(value interface{}) (string, error) {
+	v := reflect.ValueOf(value)
+	if v.Kind() == reflect.Interface {
+		v = v.Elem()
+	}
+
+	switch v.Kind() {
+	case reflect.Bool:
+		if v.Bool() {
+			return "true", nil
+		}
+		return "false", nil
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32:
+		return strconv.FormatInt(v.Int(), 10), nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32:
+		return strconv.FormatUint(v.Uint(), 10), nil
+	case reflect.Float32, reflect.Float64:
+		return strconv.FormatFloat(v.Float(), 'f', -1, 64), nil
+	case reflect.String:
+		return fmt.Sprintf("%q", v.String()), nil
```

Comment:
%q will escape like C# expects?

Reply:
Technically it will escape the way Go escapes strings, but that's very unlikely to be an issue. Though there are differences, they typically only arise in rare cases.